File: day-02/main.py
from collections import Counter
import itertools

with open('input.txt') as f:
    strs = f.read().split()

# Part 1
twos = sum(int(any(v == 2 for v in Counter(s).values())) for s in strs)
threes = sum(int(any(v == 3 for v in Counter(s).values())) for s in strs)
print(twos * threes)

# Part 2
for sa, sb in itertools.combinations(strs, 2):
    common = [a for a, b in zip(sa, sb) if a == b]
    if len(common) == len(sa) - 1:
        print(''.join(common))
        break

File: day-02/src/main.rs
use std::fs;

use utils;

fn main() {
    let input = fs::read_to_string("day-02/input.txt").ok().unwrap();  // temporary bound to the scope
    let strs: Vec<&str> = input.split_terminator("\n").collect();

    // Part 1
    let counts = strs.iter().map(|s| utils::counter(s.chars())).collect::<Vec<_>>();
    let twos = counts.iter().filter(|h| h.iter().any(|(_, &v)| v == 2)).count();
    let threes = counts.iter().filter(|h| h.iter().any(|(_, &v)| v == 3)).count();
    println!("{}", twos * threes);

    // Part 2
    let mut found = false;
    for (i, sa) in strs.iter().enumerate() {
        for sb in strs[i..].iter() {
            let common = sa.chars().zip(sb.chars()).filter(|(a, b)| a == b);
            if common.count() == sa.len() - 1 {
                let common_str = sa.chars().zip(sb.chars()).filter(|(a, b)| a == b)  // had to do it again because of ownership transfer
                        .map(|(a, _)| a.to_string()).collect::<Vec<String>>().concat();
                println!("{}", common_str);
                found = true;
                break;
            }
        }
        if found {
            break;
        }
    }
}
